File: frontend/src/types/scheduled.ts
export type PatternType = 'preset' | 'custom';

export interface ActiveDownload {
  id: number;
  episode_number: string;
  status: string;
  progress: number;
  speed_bps: number;
}

export interface ScheduledDownload {
  id: number;
  anime_id: number;
  anime_slug: string;
  anime_title: string;
  cover_url: string | null;
  source_site: string;
  dest_folder: string;
  filename_template: string;
  filename_template_type: PatternType;
  enabled: boolean;
  last_run_at: string | null;
  last_error: string | null;
  created_at: string;
  updated_at: string;
  current_episode: number;
  active_downloads: ActiveDownload[];
}

export interface ScheduleCreateRequest {
  anime_id: number;
  anime_slug: string;
  anime_title: string;
  cover_url?: string | null;
  source_site: string;
  dest_folder: string;
  filename_template: string;
  filename_template_type: PatternType;
  enabled?: boolean;
}

export interface ScheduleUpdateRequest {
  dest_folder?: string;
  filename_template?: string;
  filename_template_type?: PatternType;
  enabled?: boolean;
}

export interface ScheduleListResponse {
  scheduled: ScheduledDownload[];
  cron_expr: string;
  next_run_at: string | null;
}

export interface CronValidationResponse {
  valid: boolean;
  next_runs: string[];
  error: string | null;
}

export interface RunNowResponse {
  enqueued_episodes: number;
  skipped_reason: string | null;
}

export const PATTERN_PRESETS: { id: string; label: string; template: string }[] = [
  {
    id: 'plex_default',
    label: 'Plex (Show - S01E001 - Titolo)',
    template: '{anime} - S{season}E{episode} - {ep_title}.{ext}',
  },
  {
    id: 'episode_only',
    label: 'Solo episodio (0006.mp4)',
    template: '{episode}.{ext}',
  },
  {
    id: 'anime_and_episode',
    label: 'Anime + episodio (Show - 0006.mp4)',
    template: '{anime} - {episode}.{ext}',
  },
];

File: backend/app/schemas/scheduled.py
from datetime import datetime

from pydantic import BaseModel, Field


class ScheduleCreate(BaseModel):
    anime_id: int
    anime_slug: str
    anime_title: str
    cover_url: str | None = None
    source_site: str
    dest_folder: str = Field(..., description="Relative to /downloads")
    filename_template: str
    filename_template_type: str  # "preset" | "custom"
    enabled: bool = True


class ScheduleUpdate(BaseModel):
    dest_folder: str | None = None
    filename_template: str | None = None
    filename_template_type: str | None = None
    enabled: bool | None = None


class ScheduleResponse(BaseModel):
    id: int
    anime_id: int
    anime_slug: str
    anime_title: str
    cover_url: str | None
    source_site: str
    dest_folder: str
    filename_template: str
    filename_template_type: str
    enabled: bool
    last_run_at: datetime | None
    last_error: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ScheduleListResponse(BaseModel):
    scheduled: list[ScheduleResponse]
    cron_expr: str
    next_run_at: datetime | None


class CronValidationResponse(BaseModel):
    valid: bool
    next_runs: list[datetime] = []
    error: str | None = None


class CronUpdateRequest(BaseModel):
    cron_expr: str


class RunNowResponse(BaseModel):
    enqueued_episodes: int
    skipped_reason: str | None = None


class RunAllNowResponse(BaseModel):
    total_enqueued: int
